### PR TITLE
[Serve] [Dashboard] Add serve controller metrics to serve system dashboard page

### LINKED ISSUE #35667: [serve] Add metrics for Serve reconcile loop health
We should have a monotonically-increasing counter as well as an average latency. These should be reported under system metrics on the dashboard.

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Adds the following Serve controller metrics to the Serve details page on the dashboard:

```
ray_serve_num_ongoing_http_requests
ray_serve_num_ongoing_grpc_requests
ray_serve_controller_num_starts
ray_serve_num_scheduling_tasks
ray_serve_num_scheduling_tasks_in_backoff
ray_serve_controller_num_control_loops
```
https://github.com/ray-project/ray/pull/43797#issuecomment-1989365742 contains a screenshot.

## Related issue number

Closes https://github.com/ray-project/ray/issues/35667
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by shrekris-anyscale** [comment]
Awesome, thanks for getting this out so quickly! The dashboard page looks good.

It's a little odd that ray_serve_controller_num_starts, ray_serve_num_scheduling_tasks, and ray_serve_num_scheduling_tasks_in_backoff aren't showing up. ray_serve_controller_num_starts should show up soon after the controller initializes for the first time. ray_serve_num_scheduling_tasks and ray_serve_num_scheduling_tasks_in_backoff should show up after a few requests on a hello world app.

Is there any chance that the query is filtering for an incorrect/unexpected tag?

**Review by GeneDer — COMMENTED** [review]
I think ray_serve_num_scheduling_tasks, and ray_serve_num_scheduling_tasks_in_backoff are both gauge just happened to set to 0. There actually data in your screenshot :) 

ray_serve_controller_num_starts is indeed the odd one that doesn't come with data. Would need to dig into how come it didn't log a metrics on start. Could be a bug in the code somewhere 😕

**Comment by edoakes** [comment]
> Thanks for the reviews, I'll update the PR soon.
> 
> @edoakes any thoughts on other metrics that should be included here?

Oh nice, this is looking good! Could we please also add `serve_controller_num_control_loops`? This detects if the controller is "stuck" (counter is not increasing)

File: dashboard/client/src/pages/serve/ServeMetricsSection.tsx [hunk 1]
```diff
@@ -75,11 +75,39 @@ const METRICS_CONFIG: MetricConfig[] = [
...
+
+type ServeMetricsSectionProps = ClassNameProps & {
+  metricsConfig: MetricConfig[];
+};
 
 export const ServeMetricsSection = ({
   className,
+  metricsConfig,
 }: ServeMetricsSectionProps) => {
+  console.log(metricsConfig);
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply dashboard/client/src/pages/serve/ServeMetricsSection.tsx]: Good catch, fixed!

File: dashboard/client/src/pages/serve/ServeMetricsSection.tsx [hunk 2]
```diff
@@ -75,11 +75,39 @@ const METRICS_CONFIG: MetricConfig[] = [
   },
 ];
 
-type ServeMetricsSectionProps = ClassNameProps;
+// NOTE: please keep the titles here in sync with dashboard/modules/metrics/dashboards/serve_dashboard_panels.py
+export const SYSTEM_METRICS_CONFIG: MetricConfig[] = [
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply dashboard/client/src/pages/serve/ServeMetricsSection.tsx]: Good point, fixed

**Comment by architkulkarni** [comment]
Will remove `serve_controller_num_starts` for now since it doesn't seem like it's being exported by Serve. Created an issue here: https://github.com/ray-project/ray/issues/43829

**Comment by edoakes** [comment]
> Will remove `serve_controller_num_starts` for now since it doesn't seem like it's being exported by Serve. Created an issue here: #43829

NP. This is somewhat redundant with the control loop counter anyways

**Review by GeneDer — APPROVED** [review]
LGTM!

**Comment by architkulkarni** [comment]
Screenshot 2024-03-11 at 1 21 05 PM
Latest commit looks good, merging

**Pull request merged**